First pass at using platform abstraction for plugins

fuzzybinary opened the pull request:
Change the flutter create --template plugin command to create a PlatformInterface as an abstraction for plugins.

Work to fix #92279

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

fuzzybinary committed:
First pass at using platform abstraction for plugins

Change the flutter create --template plugin command to create a PlatformInterface as an abstraction for plugins.

fuzzybinary committed:
Rename test plugin name to avoid package conflict

dart_plugin_registry_test creates a plugin called plugin_platform_interface, which conflicts with an actual plugin now used for platform abstractions. Renamed to plugin_platform_base.

stuartmorgan commented:
> This doesn't hit some of the requests of the ticket including showing an FFI implementation

I think you misread the issue slightly. I wasn't saying that the template should demonstrate using FFI, but that using a platform interface in the template makes it easier for people developing plugins to adopt non-platform-channel implementations (such as FFI) on some platforms.

Hixie commented:
This kind of change should be tested (even though similar code currently lacks tests).

fuzzybinary committed:
Add tests that generated plugins pass analysis and unit tests

fuzzybinary commented:
I've added a test to check that we create the plugin platform interface file, and added lines that check that the generated projects and examples pass analysis and unit tests. Since the internal test uses the platform channel, I *think* this covers the functionality I'm looking to test.

christopherfujino commented:
@stuartmorgan can you take another look at this one?

christopherfujino commented:
@fuzzybinary it looks like you have merge conflicts, can you resolve them by rebasing upstream?

stuartmorgan commented:
This is near the top of my review queue BTW; sorry it had taken so long to get back to.

skia-gold commented:
Gold has detected about 59 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/92672

fuzzybinary commented:
Rebased but it looks like some tests are failing on Mac. I'll come back and take a look Monday when I have some time.

fuzzybinary commented:
Okay! Re-rebased and looks like tests are passing (no changes) so must have been an infra thing. Sorry for the delay in checking.

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl:
```diff
@@ -0,0 +1,29 @@
+  Future<String?> get platformVersion {
+    throw UnimplementedError('platformVersion() has not been implemented.');
+  }
+}
```
Please add the missing final newlines to all of these files.

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:
```diff
@@ -0,0 +1,19 @@
+import 'dart:async';
+
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
```
You shouldn't need `meta` here; `flutter/foundation.dart` vends common annotations.

(I'm not sure how you were able to compile this template, since `meta` isn't in the pubspec.)

fuzzybinary replied:
Not sure why it compiles, but meta is needed for @visibleForTesting.

stuartmorgan replied:
No it isn't; you can use `flutter/foundation.dart` as I said in my comment.

https://github.com/flutter/flutter/blob/master/packages/flutter/lib/foundation.dart#L20

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:
```diff
@@ -0,0 +1,19 @@
+  final methodChannel = const MethodChannel('{{projectName}}');
+
+  @override
+  Future<String?> get platformVersion async {
```
Same comment here about getters; use `getPlatformVersion() async`.

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:
```diff
@@ -0,0 +1,19 @@
+import 'dart:async';
```
Is this actually being used?

fuzzybinary replied:
Don't you need it for Future?

stuartmorgan replied:
Simple `Future` usage is vended by `dart:core` starting in Dart 2.1.

fuzzybinary replied:
Nope, I'm wrong about that too. Removed.

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_method_channel.dart.tmpl:
```diff
@@ -0,0 +1,19 @@
+
+  @override
+  Future<String?> get platformVersion async {
+    final version = await methodChannel.invokeMethod('getPlatformVersion');
```
Let's update this to `invokeMethod<String>`

stuartmorgan commented on packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl:
```diff
@@ -1,23 +1,27 @@
+  late Mock{{pluginDartClass}}Platform fakePlatform;
+  
   setUp(() {
-    channel.setMockMethodCallHandler((MethodCall methodCall) async {
```
All of the code you've removed from this file should still exist, just as a file called `projectName_method_channel_test.dart` (written against the platform interface though).

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl:
```diff
@@ -4,16 +4,25 @@
 class {{pluginDartClass}} {
-  static const MethodChannel _channel = MethodChannel('{{projectName}}');
+  factory {{pluginDartClass}}() {
+    _singleton ??= {{pluginDartClass}}._();
```
Why have you restructured this class? You could have kept exactly the same structure and just had the implementation call `{{pluginDartClass}}Platform.instance.platformVersion`, so this doesn't seem related to the goal of the PR.

fuzzybinary replied:
I think I was just bringing it in line with other examples I'd seen for this interface. I'll back it out and it can be a separate PR if desired.

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl:
```diff
@@ -4,16 +4,25 @@
-  static Future<String?> get platformVersion async {
-    final String? version = await _channel.invokeMethod('getPlatformVersion');
-    return version;
+  Future<String?> get platformVersion {
```
Let's change this from `get platformVersion` to `getPlatformVersion()`. Using a getter here is [against Flutter style](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#getters-feel-faster-than-methods) (not that we have to follow Flutter style in a package, but the arguments that are made there are generally applicable), and even more importantly it doesn't scale to anything with arguments, making it harder to follow as a modifiable starting point.

stuartmorgan commented on dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart:
```diff
@@ -70,7 +70,7 @@ class PluginPlatformInterfaceMacOS {
       pluginImplPubspecContent = pluginImplPubspecContent.replaceFirst(
           '    platforms:\n',
-          '    implements: plugin_platform_interface\n'
+          '    implements: plugin_platform_base\n'
```
This is a really confusing name. The better fix here would be to change `plugin` to `aplugin` or `someplugin` everywhere in this test, and keep the standard suffixes.

stuartmorgan-g reviewed (changes requested):
Sorry again for the delay. Overall this looks great, but some specific comments.

@ditman should look at the web template changes.

ditman reviewed (commented):
The web implementation looks correct. This should be easier to demonstrate by creating a `plugin` with this template and ensuring that the example app works correctly with `flutter run -d chrome` 

Also: do we want to update the `plugin` template, or create a new template for this example, @stuartmorgan ?

fuzzybinary committed:
Incorporate PR feedback

fuzzybinary commented on dev/devicelab/lib/tasks/dart_plugin_registry_tests.dart:
```diff
@@ -31,46 +31,46 @@ TaskFunction dartPluginRegistryTest({
             '--platforms',
             'macos',
-            'plugin_platform_implementation',
+            'aplugin_platform_implementation',
```
@stuartmorgan I think this is what you meant when you said use 'aplugin', can you confirm?

stuartmorgan replied:
Yes, that's correct.

stuartmorgan commented:
> Also: do we want to update the `plugin` template, or create a new template for this example, @stuartmorgan ?

We should update the existing template. This makes it easier to support web, makes it easier to support Dart-only implementations (e.g., Linux or Windows FFI), makes hybrid Dart/native platform implementations easier, and provides a much simpler path to converting a plugin to federation later. The only downside is one unnecessary API layer to plumb things though in the case where someone doesn't want any of that, but that seems like a very small price to pay. And if someone is really sure they don't want it, they can easily rip it out.)

(Also, when we later add a federated template, having this structure in the non-federated version will mean they are more similar, which I think will be helpful in the ecosystem in general.)

stuartmorgan commented:
> This should be easier to demonstrate by creating a `plugin` with this template and ensuring that the example app works correctly with `flutter run -d chrome`

Do we have any e2e web tests currently? E.g., something like the Android and iOS tests that use https://github.com/flutter/flutter/blob/0e2f51dfd04eca2935318389a75e8bc112809342/dev/devicelab/lib/tasks/plugin_tests.dart ?

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName.dart.tmpl:
```diff
@@ -1,19 +1,16 @@
-// To add platforms, run `flutter create -t plugin --platforms <platforms> .` under the same
-// directory. You can also find a detailed instruction on how to add platforms in the `pubspec.yaml` at https://flutter.dev/docs/development/packages-and-plugins/developing-packages#plugin-platforms.
+// You have generated a new plugin project without specifying the `--platforms`
+// flag. A plugin project supports no platforms is generated. To add platforms,
```
This sentence doesn't actually make sense (a pre-existing issue); as long as you're adjusting the comment let's fix it. I think it was intended to say something like "A plugin project that supports no platform was generated."

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl:
```diff
@@ -0,0 +1,27 @@
+import 'package:plugin_platform_interface/plugin_platform_interface.dart';
+import '{{projectName}}_method_channel.dart';
```
Nit: Add a blank line above this one to separate package imports from local imports.

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_platform_interface.dart.tmpl:
```diff
@@ -0,0 +1,27 @@
+  /// Defaults to [MethodChannel{{pluginDartClass}}].
+  static {{pluginDartClass}}Platform get instance => _instance;
+  
+  /// Platform-specific plugins should set this with their own platform-specific
```
Nit: s/plugins/implementations/

stuartmorgan commented on packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:
```diff
@@ -1,43 +1,25 @@
   /// Returns a [String] containing the version of the platform.
-  Future<String> getPlatformVersion() {
+  @override
+  Future<String?> getPlatformVersion() {
```
@ditman Should we make this `async` so that it can just `return value;`, or is there a downside to that?

stuartmorgan commented on packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl:
```diff
@@ -1,23 +1,27 @@
+}
 
+void main() {
+  late {{pluginDartClass}} {{pluginClassLowerCamelCase}};
```
There's no reason for the object under test to be part of the fixture, and it's a practice that tends to make tests harder to understand (and doesn't scale to testing non-trivially-constructed objects anway). You can just put:

```
{{pluginDartClass}} {{pluginClassLowerCamelCase}} = {{pluginDartClass}}();
```

at the start of each test case (in this case, all one of them).

stuartmorgan commented on packages/flutter_tools/templates/plugin/test/projectName_test.dart.tmpl:
```diff
@@ -1,23 +1,27 @@
+    {{pluginDartClass}}Platform.instance = fakePlatform;
+    {{pluginClassLowerCamelCase}} = {{pluginDartClass}}();
   });
 
```
We should add a test like this one:
https://github.com/flutter/plugins/blob/7686be7ecbe1cc049423715b5e2d897cb63a43c1/packages/file_selector/file_selector_platform_interface/test/file_selector_platform_interface_test.dart#L14-L16

as it's good to have that coverage to make sure the default behavior is as expected.

stuartmorgan commented on packages/flutter_tools/templates/template_manifest.json:
```diff
@@ -326,12 +326,16 @@
         "templates/plugin/lib/projectName.dart.tmpl",
+        "templates/plugin/lib/projectName_platform_interface.dart.tmpl",
+        "templates/plugin/lib/projectName_method_channel.dart.tmpl",
+        "templates/plugin/LICENSE.tmpl",
```
Were we just not creating this file then? Could you add a quick test for this fix as well?

fuzzybinary replied:
Looks like this might have been a mistake in the merge, as this file is now in templates/plugin_shared/LICENSE.

But, there wasn't a unit test for it either, so I added this file (and README.md) and removed this line.

stuartmorgan-g reviewed (changes requested):
A few more things, but otherwise looks good to me.

fuzzybinary committed:
Incorporate more PR feedback

stuartmorgan-g reviewed (approved):
LGTM, thanks!

@ditman Can you do the second review here since you've already looked at parts of it?

fuzzybinary commented:
@stuartmorgan I'm having trouble parsing some of the CI check failures. Do those all appear to be infrastructure issues or is there a failure I should be looking into?

ditman commented:
```
CIPD('ensure_installed') failed with: failed to resolve flutter/android/sdk/all/linux-amd64@latest (line 1): no such ref
```

I've checked in CIPD [here](https://chrome-infra-packages.appspot.com/p/flutter/android/sdk/all/linux-amd64/+/) and it seems that the `latest` Ref is gone. I don't have access to the trace logs. This might have changed in the way upstream configures its tests (I know there has been changes in the last few days about this!)

```
Excellent news! A package dependency has been removed!
  cli_util
To make sure we do not accidentally add this dependency back in the future,
please remove this packages from the allow-list in dev/bots/allowlist.dart.
Thanks!
```

This seems to have been removed ~13 days ago, [here](https://github.com/flutter/flutter/commit/0eae276e68ad43236217fa0c032ef253af0d0d70).

@fuzzybinary I think this needs another merge from master. I can get it done tomorrow when I checkout the branch and give it another review. Thanks for the contribution (and apologies for the delay!)

ditman committed:
Merge branch 'master' into feature/plugin_platform_interface_template

ditman commented on packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:
```diff
@@ -1,44 +1,26 @@
+  /// Constructs a {{pluginDartClass}}Web
+  {{pluginDartClass}}Web();
 
-  /// Handles method calls over the MethodChannel of this plugin.
-  /// Note: Check the "federated" architecture for a new way of doing this:
-  /// https://flutter.dev/go/federated-plugins
-  Future<dynamic> handleMethodCall(MethodCall call) async {
-    switch (call.method) {
-      case 'getPlatformVersion':
-        return getPlatformVersion();
-      default:
-        throw PlatformException(
-          code: 'Unimplemented',
-          details: '{{projectName}} for web doesn\'t implement \'${call.method}\'',
-        );
-    }
```
Good riddance!

ditman reviewed (approved):
This LGTM! I've also used it to generate a plugin, and the output is very cool!

<img width="959" alt="Screen Shot 2022-03-09 at 12 17 27 PM" src="https://user-images.githubusercontent.com/1255594/157540565-c17e520c-af58-4a9d-ac81-7c7411624733.png">

ditman commented:
> Do we have any e2e web tests currently? E.g., something like the Android and iOS tests that use https://github.com/flutter/flutter/blob/0e2f51dfd04eca2935318389a75e8bc112809342/dev/devicelab/lib/tasks/plugin_tests.dart ?

AFAIK: no we don't. But I can't remember when this template changed last (over a year ago?)
